### PR TITLE
Add preliminary SPAM responder verification routine

### PR DESCRIPTION
**Pull request by kuqin12**
# Preface

Please ensure you have read the [contribution docs](https://github.com/microsoft/mu/blob/master/CONTRIBUTING.md) prior
to submitting the pull request. In particular,
[pull request guidelines](https://github.com/microsoft/mu/blob/master/CONTRIBUTING.md#pull-request-best-practices).

## Description

Adding a preliminary implementation of SPAM responder function. It currently lives in supervisor, but will be moved to SPAM core once it can be launched.

For each item, place an "x" in between `[` and `]` if true. Example: `[x]`.
_(you can also check items in the GitHub UI)_

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

Tested on Q35.

## Integration Instructions

N/A

### CODE REVIEW

File: MmSupervisorPkg/Core/aux_config.spam.cfg [hunk 1]
```diff
@@ -1,5 +1,17 @@
 autogen = true
 
+[[key]]
+signature = 0x46504F4C
```
Comment (kuqin12) [inline MmSupervisorPkg/Core/aux_config.spam.cfg:4]: @Javagedes I think the way this signature is being accepted is a bit counter-intuitive as the definition of these signatures are normally formatted as 4 characters. Can we try to update the tool to see if any other options?

File: MmSupervisorPkg/Core/Relocate/Relocate.c [hunk 2]
```diff
@@ -758,11 +758,17 @@ SetupSmiEntryExit (
...
       mCetSupported = FALSE;
-      PatchInstructionX86 (mPatchCetSupported, mCetSupported, 1);
+      if (SmmCpuFeaturesGetSmiHandlerSize () == 0) {
+        PatchInstructionX86 (mPatchCetSupported, mCetSupported, 1);
+      }
     }
   } else {
     mCetSupported = FALSE;
-    PatchInstructionX86 (mPatchCetSupported, mCetSupported, 1);
+    if (SmmCpuFeaturesGetSmiHandlerSize () == 0) {
```
Comment (kuqin12) [inline MmSupervisorPkg/Core/Relocate/Relocate.c:767]: Double check...?

File: MmSupervisorPkg/Core/aux_config.spam.cfg (hunk 1 above)
Comment (kuqin12) [reply MmSupervisorPkg/Core/aux_config.spam.cfg]: Handled by https://github.com/microsoft/mu_feature_mm_supv/pull/267.

File: MmSupervisorPkg/Core/Relocate/Relocate.c (hunk 2 above)
Comment (kuqin12) [reply MmSupervisorPkg/Core/Relocate/Relocate.c:767]: Removed

File: SpamPkg/Core/Runtime/SpamResponderReport.c [hunk 3]
```diff
@@ -0,0 +1,889 @@
...
+    Status = EFI_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  if (SpamResponderData->Signature != SPAM_RESPONDER_STRUCT_SIGNATURE) {
+    Status = EFI_UNSUPPORTED;
+    goto Exit;
+  }
+
+  if (SpamResponderData->VersionMajor > SPAM_REPSONDER_STRUCT_MAJOR_VER) {
```
Comment (kenlautner) [inline SpamPkg/Core/Runtime/SpamResponderReport.c:622]: So this check seems to imply backwards compatibility.  Just want to confirm that's what you want.

**Review by kenlautner — APPROVED** [review]

File: SpamPkg/Core/Runtime/SpamResponderReport.c (hunk 3 above)
Comment (kuqin12) [reply SpamPkg/Core/Runtime/SpamResponderReport.c:622]: This basically implies that if there is a major version change, this version of spam core is unable to handle the incoming data. So we should not even try to proceed.

Comment (kuqin12) [reply SpamPkg/Core/Runtime/SpamResponderReport.c:622]: The next case implies that if the change is minor update, we may still have a chance to parse the data, although there might be some data fields unable to be processed.

File: SpamPkg/Include/Library/PeCoffLibNegative.h [hunk 4]
```diff
@@ -35,6 +35,10 @@
 #define IMAGE_VALIDATION_DATA_SIGNATURE    SIGNATURE_32 ('V', 'A', 'L', 'D')
 #define IMAGE_VALIDATION_ENTRY_SIGNATURE   SIGNATURE_32 ('E', 'N', 'T', 'R')
 
+#define KEY_SYMBOL_FW_POLICY_SIGNATURE   SIGNATURE_32 ('F', 'P', 'O', 'L')
```
Comment (kuqin12) [inline SpamPkg/Include/Library/PeCoffLibNegative.h:38]: Move these to spam responder header.

File: SpamPkg/Core/Runtime/SpamResponderReport.c [hunk 5]
```diff
@@ -0,0 +1,889 @@
+/** @file
+  MM Core Main Entry Point
+
+  Copyright (c) 2009 - 2014, Intel Corporation. All rights reserved.<BR>
+  Copyright (c) 2016 - 2018, ARM Limited. All rights reserved.<BR>
```
Comment (kuqin12) [inline SpamPkg/Core/Runtime/SpamResponderReport.c:5]: Add our copyright here.

File: SpamPkg/Include/Library/PeCoffLibNegative.h (hunk 4 above)
Comment (kuqin12) [reply SpamPkg/Include/Library/PeCoffLibNegative.h]: Moved

File: SpamPkg/Core/Runtime/SpamResponderReport.c (hunk 5 above)
Comment (kuqin12) [reply SpamPkg/Core/Runtime/SpamResponderReport.c:5]: Added.

File: SpamPkg/Core/Runtime/SpamResponderReport.c [hunk 6]
```diff
@@ -0,0 +1,890 @@
...
+    goto Exit;
+  } else {
+    Status = EFI_NOT_FOUND;
+    for (Index = 0; Index < HASH_COUNT; Index++) {
+      if (DigestList[Index].count == 0) {
+        continue;
+      }
+
+      if (DigestList[Index].digests[0].hashAlg == TPM_ALG_SHA256) {
+        if (CompareMem (DigestList[Index].digests[0].digest.sha256, (VOID*)PatchPcdGetPtr (PcdAuxBinHash), SHA256_DIGEST_SIZE) != 0) {
```
Comment (kuqin12) [inline SpamPkg/Core/Runtime/SpamResponderReport.c:695]: @Javagedes this is what i meant.

Comment (Javagedes) [reply SpamPkg/Core/Runtime/SpamResponderReport.c:695]: Awesome - I missed this it seems. Thanks for pointing it out :)

**Pull request merged**